org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/cvc_complex_type_2_4_bCodeAction.java

```diff
@@ -0,0 +1,117 @@
+/**
+ * Copyright (c) 2022 Red Hat Inc. and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v2.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v20.html
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ * Red Hat Inc. - initial API and implementation
+ */
+package org.eclipse.lemminx.extensions.contentmodel.participants.codeactions;
+
+import java.util.List;
+import java.util.regex.Pattern;
+import java.util.regex.Matcher;
+import java.util.Collection;
+import java.util.TreeSet;
+import java.text.Collator;
+
+import org.eclipse.lemminx.commons.BadLocationException;
+import org.eclipse.lemminx.commons.CodeActionFactory;
+import org.eclipse.lemminx.dom.DOMDocument;
+import org.eclipse.lemminx.services.extensions.ICodeActionParticipant;
+import org.eclipse.lemminx.services.extensions.IComponentProvider;
+import org.eclipse.lemminx.settings.SharedSettings;
+import org.eclipse.lemminx.utils.StringUtils;
+import org.eclipse.lsp4j.CodeAction;
+import org.eclipse.lsp4j.Diagnostic;
+import org.eclipse.lsp4j.Range;
+import org.eclipse.lsp4j.jsonrpc.CancelChecker;
+import org.eclipse.lsp4j.Position;
+import org.eclipse.lemminx.dom.DOMElement;
+import org.eclipse.lemminx.dom.DOMNode;
+
+/**
+ * CodeAction to insert expected child elements within an element
+ * 
+ * Given this XML where the expected child elements are not present as defined
+ * in the XSD:
+ * 
+ * <servlet></servlet>
+ * Error: 
+ * Child elements are missing from element:
+ * - servlet
+ *
+ * The following elements are expected:
+ * - description
+ * - display-name
+ * - icon
+ * - servlet-name
+ * 
+ * To fix the error, the code action will suggest inserting the expected
+ * elements inside the parent tag
+ * 
+ */
+public class cvc_complex_type_2_4_bCodeAction implements ICodeActionParticipant {
+
+	private static final Pattern ELEMENT_EXTRACTOR = Pattern
+			.compile("(?<!Child elements are missing from element:\\s)\\s-\\s([A-Za-z-_.]+)");
+
+	@Override
+	public void doCodeAction(Diagnostic diagnostic, Range range, DOMDocument document, List<CodeAction> codeActions,
+			SharedSettings sharedSettings, IComponentProvider componentProvider, CancelChecker cancelChecker) {
+
+		Range diagnosticRange = diagnostic.getRange();
+
+		try {
+			int startOffset = document.offsetAt(diagnosticRange.getStart()) + 1;
+			DOMNode node = document.findNodeAt(startOffset);
+			if (node == null || !node.isElement()) {
+				return;
+			}
+			DOMElement element = (DOMElement) node;
+			int elementOffset = element.getEndTagOpenOffset();
+			Position childElemenPosition = document.positionAt(elementOffset);
+
+			Collection<String> expectedElements = extractElements(diagnostic.getMessage());
```

Comment:
There are several problems with your current implementation:

 * you hard code the line seperator with `\n`, but you must use the line seperator from the XML document. You can do that with `String delimiter = document.getTextDocument().lineDelimiter(0);`
 * you don't take care of required attributes
 * you don't take care of required child element of an expected element.
 * you are using String.concat but it is better to use StringBuilder#append

To fix those all problems, I think we should use CMElementDeclaration with XMLGenerator like we did for completion. I suggest that you try it if it is not easy to use this strategy we will use your strategy with extract of expected elements.